app/models/concerns/concatenatable.rb

```diff
@@ -7,7 +7,8 @@ module Concatenatable
   # concatenate the values together into a single contiguous array
   def concatenate_arrays(query)
     arrays = where(query)
-    ids = arrays.pluck(:id, :array_index).sort_by(&:last).map(&:first)
-    ids.map { |id| find(id).values }.reduce([], :+)
+    arr_vals = arrays.pluck(:array_index, :values)
+    arr_vals = arr_vals.sort_by(&:first)
+    arr_vals.reduce([]) { |aggregate, arr_val|  aggregate.concat(arr_val.last) }
```

Comment:
In JavaScript, [`reduce` is > 3x slower than `for` loop](https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/#arrayreduce-vs-for-and-forof), so perhaps more primitive iteration would optimize performance here in Ruby, too.

Reply:
I think that is mainly true where the number of iterations is very large, and so saving a function call overhead is important.  Here, the number of iterations is typically <100 (a 1MM cell study will have ~100 data arrays),  so reduce does better since ruby can anticipate the array assembly.